### PR TITLE
Fix OIDC token refresh and nextUrl

### PR DESCRIPTION
**Pull request by zengyan-amazon**
*Issue #, if available:*

*Description of changes:*
* add support for OIDC token refresh
* support nextUrl in OIDC workflow

Tested manually

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: server/auth/types/openid/helper.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+export function getBaseRedirectUrl(config: SecurityPluginConfigType, core: CoreSetup): string {
+  if (config.openid?.base_redirect_url) {
+    const baseRedirectUrl = config.openid.base_redirect_url;
+    return baseRedirectUrl.endsWith('/') ? baseRedirectUrl.slice(0, -1) : baseRedirectUrl;
+  }
+
+  const host = core.http.getServerInfo().hostname;
+  const port = core.http.getServerInfo().port;
+  const protocol = core.http.getServerInfo().protocol;
```
Comment (yuxitang-amzn) [inline server/auth/types/openid/helper.ts:41]: `const {hostname: host, port, protocol} = core.http.getServerInfo();`

File: server/auth/types/openid/openid_auth.ts [hunk 2]
```diff
@@ -119,25 +127,77 @@ export class OpenIdAuthentication extends AuthenticationType {
...
+    // need to renew id token
+    if (cookie.credentials.refresh_token) {
+      try {
+        const query: any = {
+          grant_type: 'refresh_token',
+          client_id: this.config.openid?.client_id,
+          client_secret: this.config.openid?.client_secret,
+          refresh_token: cookie.credentials.refresh_token,
+        };
+        const refreshTokenResposne = await callTokenEndpoint(
```
Comment (yuxitang-amzn) [inline server/auth/types/openid/openid_auth.ts:153]: typo "Resposne"

File: server/auth/types/openid/routes.ts [hunk 3]
```diff
@@ -163,7 +184,10 @@ export class OpenIdAuthRoutes {
 
         // authHeaderValue is the bearer header, e.g. "Bearer <auth_token>"
         const token = cookie?.credentials.authHeaderValue.split(' ')[1]; // get auth token
-        const requestQueryParameters = `?post_logout_redirect_uri=${this.getBaseRedirectUrl()}/app/kibana`;
+        const requestQueryParameters = `?post_logout_redirect_uri=${getBaseRedirectUrl(
+          this.config,
+          this.core
+        )}/app/kibana`;
```
Comment (yuxitang-amzn) [inline server/auth/types/openid/routes.ts:190]: Do we need /app/kibana here? I suppose server will server auto redirect to it?

File: server/auth/types/openid/helper.ts (hunk 1 above)
Comment (zengyan-amazon) [reply server/auth/types/openid/helper.ts:41]: thanks, updated

File: server/auth/types/openid/openid_auth.ts (hunk 2 above)
Comment (zengyan-amazon) [reply server/auth/types/openid/openid_auth.ts:153]: fixed

File: server/auth/types/openid/routes.ts (hunk 3 above)
Comment (zengyan-amazon) [reply server/auth/types/openid/routes.ts:190]: you are right, it is not necessary, removed

**Pull request merged**